File: 2020/Day 1.js
const fs = require('fs');
const lines = fs.readFileSync('data.txt', 'utf8')
  .split(/[\r\n]{2}/g)
  .map(x => parseInt(x))
  .sort((a, b) => a- b);

const sum = (n) => n.reduce((acc, cur) => acc + cur, 0);
const mult = (n) => n?.reduce((acc, cur) => acc * cur, 1);

function search(numbers, target, n, acc = [], i0 = 0) {
  if (n === 0) {
    return sum(acc) === target ? [...acc] : null;
  }

  for (let i = i0; i < numbers.length && sum(acc) + numbers[i] <= target;  i++) {
    const result = search(numbers, target, n - 1, [...acc, numbers[i]], i0 = i + 1);
    if (result) {
      return result;
    }
  }
}

console.log('Part 1', mult(search(lines, 2020, 2)));
console.log('Part 2', mult(search(lines, 2020, 3)));

File: 2022/Day 1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
)

func Max(a []int, b int) []int {
	if a[0] > b {
		return a
	}

	a[0] = b
	sort.Ints(a)
	return a
}

func Sum(array []int) int {
	result := 0
	for _, v := range array {
		result += v
	}

	return result
}

func main() {
	readFile, err := os.Open("data.txt")
	if err != nil {
		fmt.Println(err)
	}

	fileScanner := bufio.NewScanner(readFile)
	fileScanner.Split(bufio.ScanLines)

	max := []int{0, 0, 0}
	cur := 0
	for fileScanner.Scan() {
		line := fileScanner.Text()
		if len(line) > 0 {
			n, _ := strconv.Atoi(line)
			cur += n
		} else {
			max = Max(max, cur)
			cur = 0
		}
	}

	readFile.Close()
	fmt.Println("Part 1:", Max(max, cur)[2])
	fmt.Println("Part 2:", Sum(Max(max, cur)))
}
